server/lib/src/plugins/namehistory.rs

```diff
@@ -0,0 +1,351 @@
+    Name,
+    LegalName,
+    DisplayName,
+}
+
+impl ToString for name_type {
+    fn to_string(&self) -> String {
+        match self {
+            name_type::Name => "name",
+            name_type::LegalName => "legalname",
+            name_type::DisplayName => "displayname",
+        }
+        .to_string()
+    }
+}
+
+impl FromStr for name_type {
+    type Err = &'static str;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        return match s {
+            "name" => Ok(name_type::Name),
+            "legalname" => Ok(name_type::LegalName),
+            "displayname" => Ok(name_type::DisplayName),
+            _ => Err("must be either: name, legalname, or displayname"),
+        };
+    }
+}
+
+lazy_static! {
+    // the naming can be confusing, here it means: 'the set of the uuids of the entries we want to target'
+    static ref entries_uuids_to_update: HashSet<PartialValue> = {
+        let mut set = HashSet::new();
+        set.insert(PartialValue::new_iutf8("account"));
+        set
+    };
+}
+
+impl NameHistory {
+    //this function returns the entries to update together with their original index they had
+    //in the cand vector. This is useful as original index is often needed to read the
+    //pre_cand vector which shares the same entries and thus the same indexes with the cand vector
+    fn entries_to_update<VALID, STATE>(
+        cand: &mut Vec<Entry<VALID, STATE>>,
+    ) -> Vec<(usize, &mut Entry<VALID, STATE>)> {
+        let mut target_cand = cand
+            .iter_mut()
+            .enumerate()
+            .filter_map(|(i, elm)| {
+                let is_target_entry = entries_uuids_to_update.iter().fold(true, |acc, current| {
+                    elm.attribute_equality("class", current) && acc
+                });
+                return if is_target_entry {
+                    Some((i, elm))
+                } else {
+                    None
+                };
+            })
+            .collect::<Vec<(usize, &mut Entry<VALID, STATE>)>>();
+        return target_cand;
+    }
+
+    fn updated_names(modification: &ModifyList<ModifyValid>) -> HashSet<name_type> {
+        let mut modified_names: HashSet<name_type> = HashSet::new();
+
+        let add_name_if_valid = |name: &str, name_set: &mut HashSet<name_type>| {
+            if let Ok(name_as_name_enum) = name_type::from_str(name) {
+                name_set.insert(name_as_name_enum);
+            };
+        };
+        modification.into_iter().for_each(|elm| {
+            match elm {
+                Modify::Present(val, _) => add_name_if_valid(val.as_str(), &mut modified_names),
+                Modify::Removed(val, _) => add_name_if_valid(val.as_str(), &mut modified_names),
+                Modify::Purged(val) => add_name_if_valid(val.as_str(), &mut modified_names),
+                Modify::Assert(_, _) => {}
+            };
+        });
+        return modified_names;
+    }
+
+    //this function is stupid and it doesn't make sense
```

Comment:
Why do you think it's stupid?

Reply:
because I think it returns the first name it finds, whereas there could be multiple previous names if we had multiple updates